### ISSUE TITLE
cross-orthogonal constraint loss and In-Batch Contrastive Learning loss

### ISSUE THREAD

**Issue by hemilu328-gif** [AIDC-AI/Marco-Voice#12]
```
    if self.orth_loss and self.emotion_embedding:
          embedding = self.speaker_projector(embedding)
          emotion_embedding = self.emotion_projector(emotion_embedding)
          embedding += emotion_embedding
          if self.cross_orth_loss:
              orth_loss = 0.0
              batch_size = embedding.size(0)
              # print("batch_size:", batch_size)
              for i in range(batch_size):
                  for j in range(i + 1, batch_size):
                      orth_loss += torch.abs(torch.dot(embedding[i], emotion_embedding[j]))
              if batch_size == 1:
                  orth_loss = 0
              else:
                  orth_loss /= (batch_size * (batch_size - 1)) / 2
          else:
              orth_loss = OrthogonalityLoss(embedding, emotion_embedding)
      else:
          orth_loss = torch.tensor(0.0).to(device) 
```
It seems that the implementation under the if self.cross_orth_loss: code block resembles In-Batch Contrastive Learning rather than a strict cross-orthogonal constraint.
If the implementation is In-Batch Contrastive Learning:
In the implementation, are contrastive learning and orthogonal loss treated as an "OR" relationship rather than an "AND" relationship? In the paper, the overall loss is defined as: L = Lₜₜₛ + λₒᵣₜₕLₒᵣₜₕ + λₒₙₜᵣₐₛₜLₒₙₜᵣₐₛₜ

**Comment by onestardao** [comment]
quick diagnosis: maps to No 8 Debugging Is a Black Box.

why: the code treats orthogonality via nested pairwise dots which ends up behaving like an in-batch contrastive block rather than enforcing a strict orthogonality constraint.

quick checks: ensure embeddings are l2 normalized, test with batch\_size = 1, and compute the gram matrix G = E · E\_emotion^T then measure off-diagonal Frobenius norm to see if the intended orthogonality term is actually shrinking.

short fixes: replace the loop with a vectorized orthogonality loss such as loss = ||G\_off||\_F^2 or add an explicit orthogonality term to the loss in addition to contrastive loss, and add unit tests for shape and numeric stability.

this is a semantic firewall style fix so no infra change required. if you want the full 16 failure modes checklist say "link please". note the tesseract.js author starred related work.
